connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+
+import re
+import time
+from contextlib import asynccontextmanager
+from datetime import datetime, timedelta
+from functools import wraps
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+
+from connectors.source import BaseDataSource
+from connectors.utils import CancellableSleeps
+
+BASE_URL = "https://slack.com/api"
+
+CURSOR = "cursor"
+RESPONSE_METADATA = "response_metadata"
+NEXT_CURSOR = "next_cursor"
+DEFAULT_RETRY_SECONDS = 30
+
+LIST_CONVERSATIONS = "list_convos"
+JOIN_CONVERSATIONS = "join_convos"
+CONVERSATION_HISTORY = "convo_history"
+CONVERSATION_REPLIES = "threads"
+LIST_USERS = "list_users"
+
+ENDPOINTS = {
+    LIST_CONVERSATIONS: "conversations.list",
+    JOIN_CONVERSATIONS: "conversations.join",
+    CONVERSATION_HISTORY: "conversations.history",
+    CONVERSATION_REPLIES: "conversations.replies",
+    LIST_USERS: "users.list",
+}
+
+# TODO list
+# Nice to haves:
+# - expand links
+#   - links:read scope needed?
+# - configure which channels to sync via deny list
+# - write an ftest
+
+
+def retryable_aiohttp_call(retries):  # TODO, make a base client class
+    # TODO: improve utils.retryable to allow custom logic
+    # that can help choose what to retry
+    def wrapper(func):
+        @wraps(func)
+        async def wrapped(*args, **kwargs):
+            retry = 1
+            while retry <= retries:
+                try:
+                    async for item in func(*args, **kwargs):
+                        yield item
+                    break
+                except Exception as e:
+                    if retry >= retries:
+                        raise e
+
+                    retry += 1
+
+        return wrapped
+
+    return wrapper
```

Comment:
I think this code is identical to the code in `utils.py::retryable`? Can we just use `@retryable` from utils?

Reply:
This is actually copied from the SPO connector. Should it also be deduped in there too?